### PR TITLE
feat: Adds conan support via a conan 2.0 recipe

### PR DESCRIPTION
**Pull request by MikeRavenelle**
- Added a conan 2.0 recipe
- Modified the CMakeLists.txt to check for presence of conan_toolchain.cmake and include libs accordingly(for building with or without conan)
- Added .md file for conan usage

I went ahead and tested the conan package on macOS and linux with example programs(using conan). I also tested the existing build procedure on macOS to make sure I did not break anything. I did no testing on Windows as I do not currently have a Windows development machine ready for use.

I would personally like to see more testing done on other users machines before a package is submitted to the conan center.

A review of my CMakeLists.txt changes also is not be a bad idea since I will admit I am not a CMake expert.

How does versioning work? I used 0.1 as a placeholder until I got feedback.

Thanks! 

## Code change checklist

- [X] I have ensured that all methods and functions are **fully documented** using doxygen style comments.
- [X] My code follows the [coding style guide](https://dpp.dev/coding-standards.html).
- [X] I tested that my change works before raising the PR.
- [X] I have ensured that I did not break any existing API calls.
- [X] I have not built my pull request using AI, a static analysis tool or similar without any human oversight.

### CODE REVIEW

**Comment by Jaskowicz1** [comment]
Can we automate the versioning to match the `${DPP_VERSION}` cmake var? This will help automate the process. If not, put it to our latest release (10.0.29) :)

**Comment by Jaskowicz1** [comment]
I've also added the "Help Wanted" label on this PR for anyone who wants to test the conan recipe (more specially on Windows). 

Like you said, we need people to test the conan recipe so asking for people to join in and test is ideal.

File: docpages/install/install-conan.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+\page install-conan Installing with Conan 2.0
```
Comment (braindigitalis) [inline docpages/install/install-conan.md:1]: doesn't putting the docs in the same pr as the cmakelists introduce a chicken and egg situation?
when we push the cmakelists to master we make it possible to get it into Conan but we also push up docs for how to install it from Conan, when it isn't available in Conan yet.

perhaps the docs should be separated out to their own pr we can merge after Conan is live.

**Review by braindigitalis — CHANGES_REQUESTED** [review]

File: docpages/install/install-conan.md (hunk 1 above)
Comment (MikeRavenelle) [reply docpages/install/install-conan.md:1]: I agree. I will remove the md file from this branch. Create a new one and PR for the doc page.

**Comment by MikeRavenelle** [comment]
Currently some issues getting it to build on Windows. I am not a windows developer so any help is appreciated. 

image

**Review by braindigitalis — APPROVED** [review]

**Comment by braindigitalis** [comment]
-1073741819 is error C0000005 which is like a segmentation fault.
It's likely something you have installed is not set up correctly or your system has a hardware issue.

**Comment by braindigitalis** [comment]
any movement on this?

File: library-conan/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        sources = self.conan_data["sources"]
+        git.clone(url=sources["url"], target=".")
+        git.checkout(commit=sources["commit"])
+
+    def generate(self):
+        deps = CMakeDeps(self)
+        deps.generate()
+        tc = CMakeToolchain(self)
+        tc.cache_variables["CONAN_EXPORTED"] = True
+        tc.cache_variables["BUILD_VOICE_SUPPORT"] = True
```
Comment (Jaskowicz1) [inline library-conan/conanfile.py:54]: Add
```py
tc.cache_variables["DPP_BUILD_TEST"] = False
tc.cache_variables["BUILD_SHARED_LIBS"] = False
```
to this and windows support will work (`DPP_BUILD_TEST` is unrelated, it's just to save space).

**Review by Jaskowicz1 — CHANGES_REQUESTED** [review]

**Comment by Jaskowicz1** [comment]
Even with the suggested changes, it appears Conan just can't find D++ (despite all the files being there and cmake loading all the variables).

I end up getting:
```
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.22631.
-- Conan: Target declared 'dpp::dpp'
CMake Error at build/cmakedeps_macros.cmake:67 (message):
  Library 'dpp' not found in package.  If 'dpp' is a system library, declare
  it with 'cpp_info.system_libs' property
Call Stack (most recent call first):
  build/dpp-Target-release.cmake:23 (conan_package_library_targets)
  build/dppTargets.cmake:24 (include)
  build/dpp-config.cmake:16 (include)
  CMakeLists.txt:10 (find_package)
```

As far as I understand, it's a Conan bug? I could be super wrong but I really don't understand.

File: library-conan/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    default_channel = "testing"
+    license = "https://github.com/brainboxdotcc/DPP/blob/master/LICENSE"
+    url = "https://github.com/brainboxdotcc/DPP"
+    description = "D++ is a lightweight and efficient library for Discord"
+    topics = ("discord")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"shared": False, "fPIC": True}
+
+    def build_requirements(self):
```
Comment (memsharded) [inline library-conan/conanfile.py:20]:
```suggestion
    def requirements(self):
```
those are regular ``requires``, recommended to go on the ``requirements()`` method

File: library-conan/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def package(self):
+        cmake = CMake(self)
+        cmake.install()
+
+    def package_info(self):
+        self.cpp_info.libs = ["dpp"]
```
Comment (memsharded) [inline library-conan/conanfile.py:69]: As commented in https://github.com/conan-io/conan/issues/15985#issuecomment-2027752183 this might need more info, as the folders generated in ``cmake.install()`` are not matching the defaults.

File: library-conan/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    license = "https://github.com/brainboxdotcc/DPP/blob/master/LICENSE"
+    url = "https://github.com/brainboxdotcc/DPP"
+    description = "D++ is a lightweight and efficient library for Discord"
+    topics = ("discord")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"shared": False, "fPIC": True}
+
+    def build_requirements(self):
+        self.requires("nlohmann_json/3.11.2", headers=True, libs=True)
```
Comment (memsharded) [inline library-conan/conanfile.py:21]: ``headers=True, lib=True`` is the default, it is recommended to drop them and use ``self.requires()``

File: library-conan/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        git.checkout(commit=sources["commit"])
+
+    def generate(self):
+        deps = CMakeDeps(self)
+        deps.generate()
+        tc = CMakeToolchain(self)
+        tc.cache_variables["CONAN_EXPORTED"] = True
+        tc.cache_variables["BUILD_VOICE_SUPPORT"] = True
+        tc.cache_variables["DPP_BUILD_TEST"] = False
+        tc.cache_variables["BUILD_SHARED_LIBS"] = False
```
Comment (memsharded) [inline library-conan/conanfile.py:56]: If it is always going to be a static library, then better drop the ``shared`` option and declare ``package_type = "static-library"``

File: library-conan/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,315 @@
+message("-- INFO: Conan detected... finding packages")
+find_package(OpenSSL REQUIRED COMPONENTS SSL Crypto)
+find_package(ZLIB REQUIRED)
+find_package(libsodium REQUIRED)
+find_package(Opus)
+find_package(Threads REQUIRED)
+find_package(Git QUIET)
+
```
Comment (memsharded) [inline library-conan/CMakeLists.txt:8]: I understand that using Conan might need some ``find_package()`` that maybe are not used when not using Conan because some other strategy like ``FetchContent`` is done, but why the big CMakeLists.txt and all the code below? In theory Conan can integrate quite cleanly, without requiring modifying CMakeLists code.

**Review by memsharded — COMMENTED** [review]
I'd suggest using the ``test_package`` functionality that can test automatically the package when created with ``conan create .``

File: library-conan/conanfile.py (hunk 6 above)
Comment (Jaskowicz1) [reply library-conan/conanfile.py:56]: Yeah this one is an odd one, we normally don't allow static but for some reason, trying to build shared results in ZLib linking twice.

I do believe we would like to build as a shared library (as this is what we do normally).

Comment (memsharded) [reply library-conan/conanfile.py:56]: It might be in Conan 1.X, that it was overlinking (propagating too much linkage requirements). In theory, Conan 2 with the right traits/package_type, Conan should avoid this overlinking, and allow shared builds to encapsulate zlib static libs as implementation detail if desired. Feel free to submit new tickets to Conan repo for further details or discussion.

Comment (Jaskowicz1) [reply library-conan/conanfile.py:56]: > It might be in Conan 1.X, that it was overlinking (propagating too much linkage requirements). In theory, Conan 2 with the right traits/package_type, Conan should avoid this overlinking, and allow shared builds to encapsulate zlib static libs as implementation detail if desired. Feel free to submit new tickets to Conan repo for further details or discussion.

Sorry for the delayed reply, the zlib issue has been a thing as long as Mike has been trying to make this PR (I mentioned it in the issue that I posted on the conan page too!). When I then tested this PR a few days ago, it was still persisting on Conan 2.2.2.

File: library-conan/CMakeLists.txt (hunk 7 above)
Comment (Jaskowicz1) [reply library-conan/CMakeLists.txt:8]: We wanted the CMakeLists.txt to be separate (see library-vcpkg), this is just something we do with VCPKG so we wanted the same format to follow.

Comment (Jaskowicz1) [reply library-conan/CMakeLists.txt:8]: It should also be noted, our main CMakeLists.txt doesn't do everything (hence why we have `library/CMakeLists.txt`, `library-vcpkg/CMakeLists.txt`, and now `library-conan/CMakeLists.txt`)

File: library-conan/conanfile.py (hunk 6 above)
Comment (memsharded) [reply library-conan/conanfile.py:56]: I think it would be great to isolate this issue, create a ticket for it in the Conan repos (hopefully with something that we can reproduce on our end) and investigate this properly.

**Comment by memsharded** [comment]
Hi @MikeRavenelle @Jaskowicz1 

If you need some other advice or help to move this forward, please let me know, I can keep reviewing and advising. Thanks!

**Comment by Jaskowicz1** [comment]
> Hi @MikeRavenelle @Jaskowicz1
> 
> If you need some other advice or help to move this forward, please let me know, I can keep reviewing and advising. Thanks!

Hi there, sorry for the late reply, cheers for the shout! I'll be looking to pick this up again very shortly so I'll give you a shout if needs be, thanks for the help so far!

**Comment by braindigitalis** [comment]
is there any progress with this?

File: library-conan/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"shared": False, "fPIC": True}
+
+    def build_requirements(self):
+        self.requires("nlohmann_json/3.11.2", headers=True, libs=True)
+        self.requires("openssl/3.1.2", headers=True, libs=True)
+        self.requires("zlib/1.3", headers=True, libs=True)
+        self.requires("libsodium/cci.20220430", headers=True, libs=True)
+        self.requires("opus/1.4", headers=True, libs=True)
```
Comment (memsharded) [inline library-conan/conanfile.py:25]:
```suggestion
        self.requires("nlohmann_json/3.11.2")
        self.requires("openssl/3.1.2")
        self.requires("zlib/1.3")
        self.requires("libsodium/cci.20220430")
        self.requires("opus/1.4")
```

**Review by braindigitalis — CHANGES_REQUESTED** [review]
all the cmakelists in here are now completely invalidated by the changes for DAVE and removal of libsodium. If we do proceed with this, it will need to be completely redone. I recommend closing this PR and making a new one.

**Comment by braindigitalis** [comment]
This is being closed in preference of a new pr which is hopefully simpler.
The conanfile in the new pr builds, and in theory is ok, but needs feedback and probably adjustment.
See PR https://github.com/brainboxdotcc/DPP/pull/1308
